vitis/hlsBs/man1/hlsWs.1
.TH hlsWs 1 "May 2026" "v1.0" "HLS Command Manual"
.SH NAME
hlsWs \- creates, removes, lists, and gives the status of an HLS workspace
.SH SYNOPSIS
.B hlsWs [OPTIONS]

.SH DESCRIPTION
.TP
.B hlsWs
Manages an HLS workspace. It can create, remove,
replace, list or get the status of the specified workspace

.SH OPTIONS
.TP
\fB\-h, \-\-help\fR
Displays this message\

.TP
\fB\-w, \-\-workspace = \fIWORKSPACE\fR
The target workspace.

.TP
\fB\-\-project = \fBHLS PROJECT FILE\fR
A project file is the standard \fBhls\fR method of capturing some of the
stable command line options in a file.

.SH Actions

.TP
\fB\-\-dry-run\fR
Only reports the actions the command would do, but does not run the command

.TP
\fB\-\-create\fR
Creates the workspace if it does not already exist

.TP
\fB\-\-list\fR
 Lists the contents of the workspace, typically HLS components + VITIS/HLS
 specific files

.TP
\fB\-\-remove\fR
Removes the workspace and all the files associated with it. Use with caution

.TP
\fB\-\-replace\fR
Replaces the workspace. Since this is effectively a --\fIremove\fR and
--\fIcreate\fR, use with caution

.TP
\fB\-\-status\fR
A brief status report giving the full path of the workspace and whether it
exists or not


.SH Project File
A project file is used to capture values of various options that could be
specified on the command line. This saves both typing and remembering what they
are.

While the project file itself may be specified on the command line via the
\fB--project=\fR options, more typically it is specified with the environment
variable \fBHLSBS_PROJECT\fR.

For the \fBhlsWs\fR command, it is convenient to place the definition of the
workspace in this file. This reduces commands to, \fIe.g.\fR
   $ hlsWs --create

.SH Suggestions on Specifying the HLS Workspace:
While the HLS workspace may be placed anywhere, it is typically placed near the
top of the project's directory tree. Two suggestions

1. Since the contents are not usually saved to the GIT repository, one should
pick a name that will be ignored by the \fI.gitignore\fR file.

2. While the workspace can be any legitimate filename, a strong suggestion is
to include the VITIS version number in the name. This allows easy changing from
one Vitis Version to another.

To make this easy, the Vitis version number can be automatically included
in the workspace name using the usual Python variable substitution pattern,
\fIe.g.\fR

   $ hlsWs --workspace=${[MY_PROJECT}/ws/{vitis-version}

This allows movement between Vitis versions, making it easy to compare one
version with another. Better yet is to include this definition in the
\fBproject file\fR.

.SH Defaults
The workspace defaults to
      <PROJECT_ROOT>/products/ws/{vitis_version}  ..or..
      <PRODUCTS_ROOT/ws/{vitis_version}


.SH Examples:
In the following assume the Vitis version is v2024.1

  $ hlsWs --create --workspace=$MY_PROJECT/ws/{vitis_version}

  Creates:  $MY_PROJECT/ws/v2024.1

  $ hlsWs --create --dry-run

  This will use the definition of the workspace found in the HLSBS project file
  to report what would have been created if --dry-run had not been specified.

  $ hlsWs  --status
  This will use the definition of the workspace found in the HLSBS project file
  to display the status of the workspace. This is useful if one has forgotten
  the name of the workspace since it is buried in the HLSBS project file.

.SH AUTHOR
Written by JJRussell


vitis/hlsBs/vitispy/hlsWs.py
# ----------------------------------------------------------------------------
# This file is part of the 'SLAC Firmware Standard Library'. It is subject to
# the license terms in the LICENSE.txt file found in the top-level directory
# of this distribution and at:
#    https://confluence.slac.stanford.edu/display/ppareg/LICENSE.html.
# No part of the 'SLAC Firmware Standard Library', including this file, may be
# copied, modified, propagated, or distributed except according to the terms
# contained in the LICENSE.txt file.
# ----------------------------------------------------------------------------

# ------------------------------------------------------------------------------
'''
    Creates/Replaces/Removes a HLS workspace

    Args:
       -project       :  Gives the default workspace
       -w, --workspace:  Specify  the workspace, overriding the default
       --create       :  Creates  the workspace
       --list         :  List the contents of the workspace
       --replace      :  Replaces the workspace, effectively --remove and --create
       --remove       :  Removes  the workspace
       --status       :  Status of the workspace
       --dry-run      :  Go through the motions but do not perform the action

   Restrictions:
      Only one of [create,replace,remove,status] may be specified

   Returns:
      Status code:
         0 = Success
         1 = Dry-Run, likely would have been successful
         2 = Dry-Run, likely would have failed
        -1 = Failure, things such as trying to create an existing workspace
'''
# ------------------------------------------------------------------------------


from vitispy.project import Project
from vitispy.workspace import Workspace
from vitispy.printer import Printer
from vitispy.dry_run import DryRun
import vitis
from vitispy.files import is_creatable
from vitispy.manpage import display_manpage
import argparse
import sys
import os
import runpy
import glob

sys.path.append(os.path.split(os.path.split(__file__)[0])[0])

# -------------------------------------------------------
# Add the VITIS HLS python paths, remove nonexistent ones
# ------------------------------------------
runpy.run_path(os.getenv('HLSBS_IMPORT_PYPATHS'),
               run_name='add_paths ' + str(sys.path))


# ------------------------------------------------------------------------------

def print_ws(title, ws, dry_run, state_msg, dry_run_msg=True):
    report = Printer(10)

    report.header(title)
    report.line("Workspace", ws)

    if dry_run and dry_run_msg:
        state_msg += " <-- would have been if not a dry-run"
    report.line("State",  state_msg)

    if dry_run:
        report.line("Dry run", "<-- Note", '*')

    report.footer()
    return
# ------------------------------------------------------------------------------


# ------------------------------------------------------------------------------
def doit():
    # ------------------------------------------------------------------------------
    # Get the command line parameters
    parser = argparse.ArgumentParser(add_help=False,
                                     fromfile_prefix_chars='@')

    Project  .add_arguments(parser)
    Workspace.add_arguments(parser)
    parser.add_argument('-h', '--help',
                        action='store_true',
                        help='Show custom help')

    action = parser.add_mutually_exclusive_group()
    action.add_argument('--create',  action='store_true', default=False)
    action.add_argument('--replace', action='store_true', default=False)
    action.add_argument('--remove',  action='store_true', default=False)
    action.add_argument('--list',    action='store_true', default=False)
    action.add_argument('--status',  action='store_true', default=False)

    parser.add_argument('--no-project',
                        dest='no_project',
                        help='Disables the use of a project file',
                        action='store_true',
                        default=False)

    DryRun.add_arguments(parser)

    args = parser.parse_args()
    if args.help:
        display_manpage(__file__)
        return 0

    needs = Project.Need.Workspace
    project_file = None if args.no_project else args.project
    project = Project(needs,
                      project_file,
                      args.root,
                      args.products_root,
                      args.workspace)
    workspace = project.workspace

    if not workspace:
        print("\nERROR: No workspace specified either\n"
              "       a) --workspace=<workspace>\n"
              "       b) specify one in an indirect file or project file\n"
              "       c) set HLS_WORKSPACE\n")
        sys.exit(-1)

    ws = Workspace.get(workspace)
    dry_run = DryRun(args.dry_run)

    exists = Workspace.exists(ws)
    if exists:
        msg = "Removed"
    else:
        msg = "Failed, does not exist "

    if args.remove:
        # ----------------------------------------------------------------------
        if exists:
            msg = "Removed"
            flag = True
        else:
            msg = "Cannot remove, does not exist"
            flag = False

        if not dry_run:
            if exists:
                Workspace.remove(ws)
            print_ws("Workspace - Removing", ws, dry_run, msg, flag)
            status = 0

        else:
            print_ws("Workspace - Removing",
                     ws,
                     dry_run,
                     msg, flag)
            status = 1

        return status
        # ----------------------------------------------------------------------

    elif args.replace:
        # ----------------------------------------------------------------------
        if exists:
            if not dry_run:
                client = vitis.create_client()
                Workspace.replace(client, ws)
                status = 0

            else:
                status = 1

            print_ws("Workspace - Replace", ws, dry_run, "Replaced")

        else:
            if not dry_run:
                client = vitis.create_client()
                Workspace.create(client, ws)
                status = 0

            else:
                status = 2

            print_ws("Workspace - Replace", ws, dry_run, "Created")

        return status
        # ----------------------------------------------------------------------

    elif args.status:
        # ----------------------------------------------------------------------
        state_msg = "Exists" if exists else "Does not exist"
        print_ws("Workspace - Status", ws, False, state_msg)
        if exists:
            status = 0
        else:
            status = -1
        return status
        # ----------------------------------------------------------------------

    elif args.list:
        # ----------------------------------------------------------------------
        printer = Printer(15)
        printer.header("Workspace - list")
        printer.line("Workspace",    ws)

        if not exists:
            printer.line("State", 'Does not Exist', '*')
            printer.footer()
            status = -1
        else:
            printer.line("State", 'Exists', ':')
            printer.separator('-')

            printer.line("Components", '')
            wc = os.path.join(ws, '*')
            cmps = glob.glob(wc)
            if len(cmps) == 0:
                printer.itemPlain("None", '')
                status = 0
                printer.footer()
                return status

            # Make a list of the stuff found and the longest one
            cmp_names = []
            max = 0
            for cmp in cmps:
                cmp_name = os.path.split(cmp)[1]
                cmp_names.append(cmp_name)
                n = len(cmp_name)
                if n > max:
                    max = n
            cmp_names.sort()

            # -------------------------------------------------------------
            # Fill a line until it can no longer completely hold a new entry
            # --------------------------------------------------------------
            per_line = 80 // max
            left = per_line
            line = ''
            cnt = len(cmps)
            for cmp_name in cmp_names:
                line += f"{cmp_name:{max}s} "
                cnt -= 1
                left -= 1
                if left <= 0 or cnt == 0:
                    print(line)
                    if cnt == 0:
                        break
                    line = ''
                    left = per_line

            printer.footer()
            status = 0

        return status
        # ----------------------------------------------------------------------

    elif args.create:
        # ----------------------------------------------------------------------
        if not exists:
            # --------------------------------------
            # Check if the directory can be created
            # The vitis code gives an obscure error if the path is not valid
            # --------------------------------------------------------------
            check = is_creatable(ws)

            if not check:
                print_ws("Workspace - Create", ws, dry_run, check,
                         "Failed, the workspace is an invalid path", False)
                if dry_run:
                    status = 2
                else:
                    status = 1

            elif not dry_run:
                client = vitis.create_client()
                Workspace.create(client, ws)

                print_ws("Workspace - Create", ws, dry_run, "Created")
            else:
                print_ws("Workspace - Create", ws, dry_run, "Created")

            if dry_run:
                status = 1
            else:
                status = 0

        else:

            print_ws("Workspace - Create",
                     ws,
                     dry_run,
                     "Already Exists, use --replace", False)
            if dry_run:
                status = 2
            else:
                status = 1

        return status
        # ----------------------------------------------------------------------

    else:
        # ----------------------------------------------------------------------
        print(
            "WARNING: No action was specified. Select from \n"
            "         --help, --create, --replace, --remove, --list or --status",
            file=sys.stderr)
        status = -1
        return status
    # --------------------------------------------------------------------------


# ------------------------------------------------------------------------------


if __name__ == '__main__':
    status = doit()
    sys.exit(status)
